windows/flutter_nsd_plugin.cpp

```diff
@@ -0,0 +1,335 @@
+      const flutter::MethodCall<flutter::EncodableValue> &method_call,
+      std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result);
+  std::shared_ptr<flutter::MethodChannel<flutter::EncodableValue>> channel;
+
+  MdnsRequest* currentRequest;
+
+};
+
+// static
+void FlutterNsdPlugin::RegisterWithRegistrar(
+    flutter::PluginRegistrarWindows *registrar) {
+  auto channel =
+      std::make_shared<flutter::MethodChannel<flutter::EncodableValue>>(
+          registrar->messenger(), "com.nimroddayan/flutter_nsd",
+          &flutter::StandardMethodCodec::GetInstance());
+
+  auto plugin = std::make_unique<FlutterNsdPlugin>(channel);
+
+  channel->SetMethodCallHandler(
+      [plugin_pointer = plugin.get()](const auto &call, auto result) {
+        plugin_pointer->HandleMethodCall(call, std::move(result));
+      });
+
+  registrar->AddPlugin(std::move(plugin));
+}
+
+
+
+FlutterNsdPlugin::FlutterNsdPlugin(std::shared_ptr<flutter::MethodChannel<flutter::EncodableValue>> _channel)
+
+{
+  channel = _channel;
+  currentRequest = NULL;
+
+  WORD versionWanted = MAKEWORD(1, 1);
+  WSADATA wsaData;
+  if (WSAStartup(versionWanted, &wsaData)) {
+
+    printf("Failed to initialize WinSock\n");
+  }
+}
+
+FlutterNsdPlugin::~FlutterNsdPlugin() {
+  WSACleanup();
+}
+
+
+
+void FlutterNsdPlugin::HandleMethodCall(
+  const flutter::MethodCall<flutter::EncodableValue>& method_call,
+  std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result) {
+  if (method_call.method_name().compare("getPlatformVersion") == 0) {
+    std::ostringstream version_stream;
+    version_stream << "Windows ";
+    if (IsWindows10OrGreater()) {
+      version_stream << "10+";
+    }
+    else if (IsWindows8OrGreater()) {
+      version_stream << "8";
+    }
+    else if (IsWindows7OrGreater()) {
+      version_stream << "7";
+    }
+    result->Success(flutter::EncodableValue(version_stream.str()));
+  }
+  else if (method_call.method_name().compare("startDiscovery") == 0) {
+    const auto* arguments = std::get_if<flutter::EncodableMap>(method_call.arguments());
+    auto serviceType = arguments->find(flutter::EncodableValue("serviceType"));
+    std::string type = "";
+    if (serviceType != arguments->end()) {
+      type = std::get<std::string>(serviceType->second);
+    }
+    
+    if (currentRequest != NULL) {
+      auto req = currentRequest;
+      currentRequest = NULL; // request will delete itself after the thread exits, we can detach the pointer here
+      req->stop(); // will stop the background thread after the next timeout; returns immediately
+    }
+    currentRequest = new MdnsRequest(channel);
+    currentRequest->start(type);
+    result->Success();
+  }
+  else if (method_call.method_name().compare("stopDiscovery") == 0) {
+    if (currentRequest != NULL) {
+      auto req = currentRequest;
+      currentRequest = NULL;
+      req->stop();
+
+    }
+    result->Success();
+  }
+  else {
+    result->Success();
```

Comment:
I'd prefer to return a failure here in case an unsupported method was called

Reply:
ugh. copy/paste error